src/eip712/transaction_request.rs

```diff
@@ -181,30 +184,31 @@ impl Eip712TransactionRequest {
         stream.append(&self.value);
         // 6
         stream.append(&self.data.0);
-        if let Some(signature) = signature {
-            // 7
-            stream.append(&U64::from(signature.v));
-            // 8
-            stream.append(&signature.r);
-            // 9
-            stream.append(&signature.s);
-        } else {
-            // 7, 8, 9 must be set even if no signature is provided.
-            // This should be the case of transaction that have a
-            // custom signature set.
-            stream.append(&"");
-            stream.append(&"");
-            stream.append(&"");
-        }
+        // 7
+        stream.append(&self.chain_id);
+        // 8
+        stream.append(&"");
+        // 9
+        stream.append(&"");
         // 10
         stream.append(&self.chain_id);
         // 11
         stream.append(&self.from);
         // 12, 13, 14, 15
-        self.custom_data.rlp_append(&mut stream);
-
+        if self.custom_data.custom_signature.clone().is_some() {
+            self.custom_data.rlp_append(&mut stream);
+        } else if let Some(signature) = signature {
+            let tx = self.clone().custom_data(
+                self.clone()
+                    .custom_data
+                    .custom_signature(signature.to_vec()),
+            );
+            tx.custom_data.rlp_append(&mut stream);
+        } else {
+            return Err(Eip712Error::Message("No signature provided".to_owned()));
```

Comment:
Is it a good idea to fail when the `signature` parameter equals `None` when to its type is `Option<Signature>`?

Reply:
It fails only when both `custom_signature` and `signature` are `None`, as at least one of them must be present.

Reply:
Ok. We should discuss in the future if it makes sense to leverage the type system to prevent an invalid state.

Reply:
I created a ticket just to track the discussion https://github.com/lambdaclass/zksync-web3-rs/issues/101.

Reply:
I'm marking this chat as resolved.